Automattic/jetpack
# Backward Compatibility Package: Proof of Concept

#### Changes proposed in this Pull Request:
Introducing the new package `jetpack-back` (working title).
It lets you ensure that a feature is supported by all the plugins that may be involved, and safely introduce new backward-incompatible changes into other packages.

#### Jetpack product discussion
tbd

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
1. Checkout the branch, run `composer install` to get the package installed.
2. Install the testing plugin:
<details>
<summary>The Testing Plugin</summary>
<p>

```php
<?php
/** Plugin Name: Jetpack Back Test */

use Automattic\Jetpack\Back;
use Automattic\Jetpack\Config;

if ( empty( $_GET['back-debug'] ) ) {
	return;
}

add_action(
	'plugins_loaded',
	function() {
		// This plugin supports latest features.
		( new Config() )->ensure(
			'back',
			array(
				'slug'     => 'sample-plugin-1',
				'features' => array(
					Back\Features::SITE_REGISTRATION_TOS_REQUIRED,
					Back\Features::SAMPLE_FEATURE_1,
				)
			)
		);

		// This plugin only supports one feature.
		( new Config() )->ensure(
			'back',
			array(
				'slug'     => 'sample-plugin-2',
				'features' => array(
					Back\Features::SITE_REGISTRATION_TOS_REQUIRED,
				)
			)
		);

		// This plugin doesn't support any features, but the `jetpack-back` package knows about it.
		/*( new Config() )->ensure(
			'connection',
			array(
				'slug' => 'sample-plugin-3',
				'name' => 'Sample Plugin 3',
			)
		);*/
	},
	1
);

add_action(
	'init',
	function() {
		var_dump(
			Back\Compatibility::can_use( Back\Features::SITE_REGISTRATION_TOS_REQUIRED ),
			Back\Compatibility::can_use( Back\Features::SAMPLE_FEATURE_1 ),
			Back\Compatibility::can_use( Back\Features::SAMPLE_FEATURE_2 )
		);
		die();
	}
);

```
</p>
</details>

3. Load the page `/wp-admin/index.php?back-debug=1` and confirm that feature `SITE_REGISTRATION_TOS_REQUIRED` is available to use (the function returns `true`), and the both "sample features" are not available (`false` is returned).
4. Uncomment the `ensure('connection', ... )` code in the plugin, reload the `back-debug` page. Make sure that all three features are not available now (`false` is returned).

#### Proposed changelog entry for your changes:
* New package to improve backward compatibility.
